### DISCUSSION TITLE
✅ Separate positive and negative response types in client

### DISCUSSION THREAD

**Discussion by shawncarr** [RobinTail/express-zod-api#1411] [Ideas]
Right now you provide a `Response` interface that looks something like

```
export interface Response extends Record<MethodPath, any> {
,,,
  'get /companies/:company': GetCompaniesCompanyResponse;
...
}
```

But to get the typing for the Positive Status of that response I have to do something like 

```
export type GetCompaniesCompanyResponsePositive = Extract<
  Response['get /companies/:company'],
  { status: 'success' }
>;
```
Then repeat for the Negative status for each response. For each client that it would be imported into.

IMHO it would be nice if you generate and return a `Responses` type something like

```
export type Responses = {
...
  GetCompaniesCompanyResponse: GetCompaniesCompanyResponse;
  GetCompaniesCompanyResponseNegative: Extract<GetCompaniesCompanyResponse, { status: 'error' }>;
  GetCompaniesCompanyResponsePositive: Extract<GetCompaniesCompanyResponse, { status: 'success' }>;
....
};
```

I know it wouldn't be that simple as this would only work for the `defaultResultHandler` as the status might not be there in a different result handler. 

Thoughts?

**Comment by RobinTail** [comment]
> But to get the typing for the Positive Status of that response I have to

You don't have to do that.
Those types are supposed to describe the _possible_ response, which can be positive OR negative. 
It is the responsibility of the client to handle and differentiate it.
Basically all you need to do in your implementation is to throw in case of the negative response detected:

```tsx
async function getSomethingFromAPI() {
  const response = await client.provide(/*...*/);
  if (response.status === "error") {
      throw new Error(response.error.message);
  }
  return response.data;
}
```

This way Typescript will automatically narrow down the return type of your function to the positive response only, while the case of the negative one is also handled, @shawncarr

**Comment by RobinTail** [comment]
> it would be nice 

Perhaps I don't understand your use case. Please provide more reasoning if the code sample above does not solve your problem.

> this would only work for the defaultResultHandler 

Those types are actually stored separately and combined intentionally into one, when generating the client.
So, technically it's easy to provide them separately, but I don't understand why, @shawncarr

**Comment by RobinTail** [comment]
Check out this sample frontend API consumer, with a variety of usage examples, @shawncarr : 

https://github.com/RobinTail/incaseofmydeath/blob/master/frontend/src/api.ts

**Comment by shawncarr** [comment]
Your example does work in a project does not require explicit return types. But if you need explicit return types then you end up with something like 
```
async function getSomethingFromAPI(): Promise<Extract<Response['get /companies/:company'],{ status: 'success' }>;> {
  const response = await client.provide(/*...*/);
  if (response.status === "error") {
      throw new Error(response.error.message);
  }
  return response.data;
}
```

Which I usually do a type to shorten it up and have re-usability 
```
export type GetCompaniesCompanyResponsePositive = Extract<
  Response['get /companies/:company'],
  { status: 'success' }
>;````

**Reply by RobinTail** [reply]
Now I understand, @shawncarr .
Thank you for the explanation. 
Is that requirement of the explicit return types a some kind of ESLint rule in the project you work on?
Anyway, I'll take a look on how to improve that and provide the response cases separately (will take some time).

**Comment by RobinTail** [comment]
Check this out, @shawncarr : 
https://github.com/RobinTail/express-zod-api/pull/1412
Would this fulfil your needs?

**Comment by shawncarr** [comment]
I like it

**Reply by RobinTail** [reply]
Great. This will be a feature in the next major release, available for you by activating the `splitResponse` option upon creation of the `Integration` instance, @shawncarr 

```ts
new Integration({ routing, splitResponse: true })
```

**Comment by RobinTail** [comment]
@shawncarr , 🚀 

I've just published dev preview version `16.0.0-beta5` having the feature you need.
It would be great if you could try it and let me know your impression.

**Comment by shawncarr** [comment]
I believe it is working as expected.
